storage/innobase/handler/ha_innodb.cc

```diff
@@ -10478,67 +10487,91 @@ ha_innobase::wsrep_append_keys(
 			KEY*  key_info	= table->key_info + i;
 			if (key_info->flags & HA_NOSAME) {
 				hasPK = true;
+				break;
 			}
 		}
 
 		for (i=0; i<table->s->keys; ++i) {
-			uint  len;
-			char  keyval0[WSREP_MAX_SUPPORTED_KEY_LENGTH+1] = {'\0'};
-			char  keyval1[WSREP_MAX_SUPPORTED_KEY_LENGTH+1] = {'\0'};
-			char* key0 		= &keyval0[1];
-			char* key1 		= &keyval1[1];
 			KEY*  key_info	= table->key_info + i;
-			ibool is_null;
 
-			dict_index_t* idx  = innobase_get_index(i);
-			dict_table_t* tab  = (idx) ? idx->table : NULL;
+			dict_index_t* idx = innobase_get_index(i);
+			dict_table_t* tab = (idx) ? idx->table : NULL;
 
+			/* keyval[] shall contain an ordinal number at byte 0
+			   and the actual key data shall be written at byte 1.
+			   Hence the total data length is the key length + 1 */
+			char keyval0[WSREP_MAX_SUPPORTED_KEY_LENGTH+1] = {'\0'};
+			char keyval1[WSREP_MAX_SUPPORTED_KEY_LENGTH+1] = {'\0'};
 			keyval0[0] = (char)i;
 			keyval1[0] = (char)i;
+			char* key0 = &keyval0[1];
+			char* key1 = &keyval1[1];
 
 			if (!tab) {
 				WSREP_WARN("MariaDB-InnoDB key mismatch %s %s",
 					   table->s->table_name.str,
 					   key_info->name.str);
 			}
-			/* !hasPK == table with no PK, must append all non-unique keys */
+			/* !hasPK == table with no PK,
+			   must append all non-unique keys */
 			if (!hasPK || key_info->flags & HA_NOSAME ||
 			    ((tab &&
 			      referenced_by_foreign_key2(tab, idx)) ||
 			     (!tab && referenced_by_foreign_key()))) {
 
-				len = wsrep_store_key_val_for_row(
+				ibool is_null0;
+				uint len0 = wsrep_store_key_val_for_row(
 					thd, table, i, key0,
 					WSREP_MAX_SUPPORTED_KEY_LENGTH,
-					record0, &is_null);
-				if (!is_null) {
-					rcode = wsrep_append_key(
-						thd, trx, table_share,
-						keyval0, len+1, key_type);
-
-					if (rcode) DBUG_RETURN(rcode);
-
-					if (key_info->flags & HA_NOSAME
-					    || key_type == WSREP_KEY_SHARED)
-						key_appended = true;
-				} else {
-					WSREP_DEBUG("NULL key skipped: %s",
-						    wsrep_thd_query(thd));
-				}
+					record0, &is_null0);
 
 				if (record1) {
-					len = wsrep_store_key_val_for_row(
+					ibool is_null1;
+					uint len1 = wsrep_store_key_val_for_row(
 						thd, table, i, key1,
 						WSREP_MAX_SUPPORTED_KEY_LENGTH,
-						record1, &is_null);
+						record1, &is_null1);
+
+					if (is_null0 != is_null1 ||
+					    len0 != len1 ||
+					    memcmp(key0, key1, len0)) {
+						/* This key has chaged. If it
```

Comment:
How does this work with collations that map multiple binary representations to the same key? Say, case-insensitive or accent-insensitive collations? Shouldn’t this use one of the member functions of `CHARSET_INFO::coll`?

Reply:
I'm not 100% sure how this works with collations, but again this is a direct copy from line 10535 in the original code so no functionality was changed. I suppose we are dealing with "collation-adjusted" key data at this point (as we should) otherwise nothing would work.